### ISSUE TITLE
Feature complete / Robust of distributed training

### ISSUE THREAD

**Issue by typhoonzero** [PaddlePaddle/Paddle#12163]
### Features

- [ ] incremental training: run `load_params` on one of the trainer and continue training @seiriosPlus
- [x] fetch initialized parameters from pserver when trainers run `startup_program` -- @velconia Doing
- [ ] support training using MPI communication APIs (in contrib) -- @seiriosPlus 
- [ ] NCCL2 distributed mode support run different number of batches on each node -- @typhoonzero
- [ ] polish NCCL2 dist training API for user.
- [ ] Using trainer_id mapping on pservers to avoid duplicated barrier messages -- @Yancey1989 
- [ ] Update transpilers to IR passes. 

### Performance-boosting:

- [ ] merge small tensors using a continuous piece of memory and send them using one RPC call. -- @velconia
- [ ] reduce copies when doing distributed sparse training. -- @Yancey1989 
- [ ] support run multiple models in trainer in async training.

### Robust:

- [ ] add a base class to set up a dist train job easier in unit test. -- @Yancey1989 #12258 
- [x] add uni test for all supported model (SE_ResNeXt50_32x4d, transformer, etc.)
- [ ] test for 1pserver and 1trainer case
- [x] test for 2pservers and 2trainers case
- [x] test for the sparse train case
- [x] test for async training case
- [ ] test for training with learning rate decay, L2 regularization, gradient clipping
- [ ] test for training with nccl2 with 2workers (TCP)

**Comment by Yancey0623** [comment]
The sparse update needs a uniform design with dist lookup table.

**Comment by seiriosPlus** [comment]
I will add a test case for ```text_classification```:
 test for 1pserver and 1trainer case
 test for 2pservers and 2trainers case

I will add a test case for ``` strict accuracy alignment ```.
